### ISSUE TITLE
DDP Freezes w/ No Output for PyTorch Geometric GNN Multi-GPU Node Classification

### ISSUE THREAD

**Issue by puririshi98** [pytorch/pytorch#79049]
### 🐛 Describe the bug

For some mysterious reason self.reducer._rebuild_buckets() from within the DDP forward call, hangs on the second iteration of calling the DDP module forward pass. It only happens on node property prediction for PyTorch Geometric, link prediction works with no issues. Originally the issue was found when predicting missing information in the Microsoft Academic Graph, but the issue has been isolated down to be reproducible with node classification on synthetic data. [Repro](https://github.com/puririshi98/debug_MAG_pyg/blob/main/synthetic_repro.py)
The lack of output when it freezes is making the debugging impossible for @rusty1s  as well.
Hacky fix:
```
            try:
                bool2 = self.reducer._rebuild_buckets()
            except:
                bool2 = False
            if torch.is_grad_enabled() and bool2:
                logger.info("Reducer buckets have been rebuilt in this iteration.")
                self._has_rebuilt_buckets = True
```
as opposed to:
```
            if torch.is_grad_enabled() and self.reducer._rebuild_buckets(:
                logger.info("Reducer buckets have been rebuilt in this iteration.")
                self._has_rebuilt_buckets = True
```

Not sure the cause or what the correct long term solution is.

### Versions

latest pytorch and pytorch-geometric installs should reproduce the issue

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501

**Comment by puririshi98** [comment]
@rohan-varma any ideas on this?

**Comment by rohan-varma** [comment]
@puririshi98 Thanks for filing this issue! I'm curious as to why switching over to try/except fixes the issue if the issue is indeed a hang (it should hang in both cases)? 

If try/except does work, could you let us know the exception that is caught here? 

Also, setting `TORCH_DISTRIBUTED_DEBUG=DETAIL` in your application can help with catching any mismatched collectives / hang issues. To debug further, could you provide a full / minimal repro of your issue to see if we can debug it on our end? Thank you!

**Comment by puririshi98** [comment]
> Also, setting TORCH_DISTRIBUTED_DEBUG=DETAIL in your application can help with catching any mismatched collectives / hang issues.

I have tried this already and unfortunately nothing comes out. The full/minimal repro was including in the original bug report.

> but the issue has been isolated down to be reproducible with node classification on synthetic data. [Repro](https://github.com/puririshi98/debug_MAG_pyg/blob/main/synthetic_repro.py)

**Comment by puririshi98** [comment]
The exception that comes out is:
```
Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by
making sure all `forward` function outputs participate in calculating loss.
If you already have done the above, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).
Parameters which did not receive grad for rank 1: layers.1.conv.convs.paper__rev_writes__author.lin_r.weight, layers.1.conv.convs.paper__rev_writes__author.lin_l.bias, layers.1.conv.convs.paper__rev_writes__author.lin_l.weight, layers.1.conv.convs.institution__rev_affiliated_with__author.lin_r.weight, layers.1.conv.convs.institution__rev_affiliated_with__author.lin_l.bias, layers.1.conv.convs.institution__rev_affiliated_with__author.lin_l.weight, layers.1.conv.convs.paper__has_topic__field_of_study.lin_r.weight, layers.1.conv.convs.paper__has_topic__field_of_study.lin_l.bias, layers.1.conv.convs.paper__has_topic__field_of_study.lin_l.weight, layers.0.conv.convs.author__affiliated_with__institution.lin_r.weight, layers.0.conv.convs.author__affiliated_with__institution.lin_l.bias, layers.0.conv.convs.author__affiliated_with__institution.lin_l.weight, layers.1.conv.convs.author__affiliated_with__institution.lin_l.weight, layers.1.conv.convs.author__affiliated_with__institution.lin_l.bias, layers.1.conv.convs.author__affiliated_with__institution.lin_r.weight
Parameter indices which did not receive grad for rank 1: 0 1 2 24 25 26 33 34 35 36 37 38 39 40 41
```
despite this ugly exception, training goes through fine and  evaluation metrics are as expected

**Comment by rohan-varma** [comment]
I see, so it seems that skipping the bucket rebuild eliminates the issue. @zhaojuanmao, do you know why bucket rebuild can result in the exception above? Thanks!

**Comment by zhaojuanmao** [comment]
This seems not to be related to '_rebuild_buckets()' algorithms, the exception is from 'ensure_prior_reduction_finished();' inside '_rebuild_buckets()'

This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by
making sure all `forward` function outputs participate in calculating loss.

So would you please try 'find_unused_parameters=True' to `torch.nn.parallel.DistributedDataParallel`?

**Comment by puririshi98** [comment]
> So would you please try 'find_unused_parameters=True' to torch.nn.parallel.DistributedDataParallel?
my original code for the MAG graph dataset had this on by default and that is how i first discovered the bug. I just added it to the synthetic data repro w/ this [commit](https://github.com/puririshi98/debug_MAG_pyg/commit/f48e5386a58e332e74b92bb120db321db532533f). I reconfirmed that this does not fix things and the bug behaves exactly as it did before

**Comment by puririshi98** [comment]
any updates on this? we have a patch working internally for a while now but this should be available to those not using our containers
[patch_ddp.zip](https://github.com/pytorch/pytorch/files/9815175/patch_ddp.zip)

**Comment by puririshi98** [comment]
any updates on this? we have a patch working internally for a while now but this should be available to those not using our containers
[patch_ddp.zip](https://github.com/pytorch/pytorch/files/9815175/patch_ddp.zip)

**Comment by puririshi98** [comment]
@rohan-varma, this issue persists even with the latest software stack. we have implemented my hacky fix as a patch in our containers, but this issue would exist for any users working with PyG outside of our container
